src/module/functions/mod.rs

```diff
@@ -447,98 +447,112 @@ impl Module {
-    ///
-    /// module.replace_exported_func(fid, move || {
+    /// module.replace_exported_func(fid, |body| {
     ///     builder.func_body().unreachable();
-    ///     builder.local_func(vec![])
     /// });
     /// ```
     ///
+    /// The arguments passed to the original function will be passed to the
+    /// new exported function that was built in your closure.
+    ///
     /// This function returns the function ID of the *new* function,
     /// after it has been inserted into the module as an export.
-    pub fn replace_exported_func<F>(&mut self, fid: FunctionId, fn_builder: F) -> Result<FunctionId>
-    where
-        F: FnOnce((&FuncParams, &FuncResults)) -> Result<LocalFunction>,
-    {
-        match (self.exports.get_exported_func(fid), self.funcs.get(fid)) {
-            (
-                Some(exported_fn),
-                Function {
-                    kind: FunctionKind::Local(lf),
-                    ..
-                },
-            ) => {
-                // Retrieve the params & result types for the exported (local) function
-                let ty = self.types.get(lf.ty());
-                let (params, results) = (ty.params().to_vec(), ty.results().to_vec());
-
-                // Add the function produced by `fn_builder` as a local function,
-                let new_fid = self.funcs.add_local(
-                    fn_builder((&params, &results)).context("export fn builder failed")?,
-                );
-
-                // Mutate the existing export to use the new local function
-                let export = self.exports.get_mut(exported_fn.id());
-                export.item = ExportItem::Function(new_fid);
-
-                Ok(new_fid)
-            }
-            // The export didn't exist, or the function isn't the kind we expect
-            _ => bail!("cannot replace function [{fid:?}], it is not an exported function"),
+    pub fn replace_exported_func(
+        &mut self,
+        fid: FunctionId,
+        builder_fn: impl FnOnce(&mut InstrSeqBuilder),
+    ) -> Result<FunctionId> {
+        let original_export_id = self
+            .exports
+            .get_exported_func(fid)
+            .map(|e| e.id())
+            .with_context(|| format!("no exported function with ID [{fid:?}]"))?;
+
+        if let Function {
+            kind: FunctionKind::Local(lf),
+            ..
+        } = self.funcs.get(fid)
+        {
+            // Retrieve the params & result types for the exported (local) function
+            let ty = self.types.get(lf.ty());
+            let (params, results) = (ty.params().to_vec(), ty.results().to_vec());
+
+            // Add the function produced by `fn_builder` as a local function
+            let mut builder = FunctionBuilder::new(&mut self.types, &params, &results);
+            let mut new_fn_body = builder.func_body();
+            builder_fn(&mut new_fn_body);
+            let func = builder.local_func(lf.args.clone());
```

Comment:
`lf.args()` is the local list I'm referring to fwiw.

Reply:
Ah thanks this is helpful -- I'll pass it through from here